File: apps/web/src/modules/contacts/vcard.ts
// vCard 3.0 / 4.0 parse + emit (plan §3 e7, §2.1). A small hand-rolled reader/
// writer that maps the frozen `ContactCard` shape ⇄ vCard bytes for the
// import/export UI. The engine's `mw-ics` (`vcard4` crate) is the round-trip
// source of truth at integration (e10); this client-side pass powers the import
// *preview* + a local export without a server round-trip, so it stays lenient:
// unknown properties are ignored, both 3.0 and 4.0 spellings are accepted.

import type {
  Anniversary,
  ContactCard,
  ContactEmail,
  ContactName,
  ContactValue,
} from '../../api/pim-types.ts';

/** A parsed contact plus a stable client id (import preview needs a key). */
export type ParsedContact = Omit<ContactCard, 'id' | 'addressBookId' | 'etag'>;

interface VCardProperty {
  name: string;
  params: Record<string, string[]>;
  value: string;
}

// ── Reading ──────────────────────────────────────────────────────────────────

/** Unfold RFC 6350 line folding: a leading space/tab continues the prior line. */
function unfold(text: string): string[] {
  const raw = text.replace(/\r\n/g, '\n').replace(/\r/g, '\n').split('\n');
  const out: string[] = [];
  for (const line of raw) {
    if ((line.startsWith(' ') || line.startsWith('\t')) && out.length > 0) {
      out[out.length - 1] += line.slice(1);
    } else {
      out.push(line);
    }
  }
  return out;
}

/** Split a string on `sep`, honouring double-quoted spans (for params). */
function splitQuoted(input: string, sep: string): string[] {
  const parts: string[] = [];
  let cur = '';
  let quoted = false;
  for (const ch of input) {
    if (ch === '"') {
      quoted = !quoted;
      cur += ch;
    } else if (ch === sep && !quoted) {
      parts.push(cur);
      cur = '';
    } else {
      cur += ch;
    }
  }
  parts.push(cur);
  return parts;
}

/** Unescape a vCard TEXT value (`\n` `\,` `\;` `\\`). */
function unescapeText(value: string): string {
  let out = '';
  for (let i = 0; i < value.length; i += 1) {
    const ch = value[i];
    if (ch === '\\' && i + 1 < value.length) {
      const next = value[i + 1]!;
      if (next === 'n' || next === 'N') out += '\n';
      else out += next;
      i += 1;
    } else {
      out += ch!;
    }
  }
  return out;
}

/** Split a structured/list value on unescaped `sep`. */
function splitValue(value: string, sep: string): string[] {
  const parts: string[] = [];
  let cur = '';
  for (let i = 0; i < value.length; i += 1) {
    const ch = value[i]!;
    if (ch === '\\' && i + 1 < value.length) {
      cur += ch + value[i + 1]!;
      i += 1;
    } else if (ch === sep) {
      parts.push(cur);
      cur = '';
    } else {
      cur += ch;
    }
  }
  parts.push(cur);
  return parts;
}

function parseLine(line: string): VCardProperty | null {
  // Split name+params from value at the first unquoted colon.
  const quoted = splitQuoted(line, ':');
  if (quoted.length < 2) return null;
  const head = quoted[0]!;
  const value = quoted.slice(1).join(':');
  const segs = splitQuoted(head, ';');
  let name = segs[0]!;
  // Drop any grouping prefix (`item1.EMAIL` → `EMAIL`).
  const dot = name.indexOf('.');
  if (dot >= 0) name = name.slice(dot + 1);
  name = name.trim().toUpperCase();
  const params: Record<string, string[]> = {};
  for (const seg of segs.slice(1)) {
    const eq = seg.indexOf('=');
    if (eq < 0) {
      // A bare `TYPE`-less param value (vCard 2.1 style): treat as a TYPE.
      const bare = seg.replace(/"/g, '').trim().toUpperCase();
      if (bare.length > 0) (params['TYPE'] ??= []).push(bare);
      continue;
    }
    const key = seg.slice(0, eq).trim().toUpperCase();
    const vals = splitQuoted(seg.slice(eq + 1), ',').map((v) => v.replace(/"/g, '').trim());
    (params[key] ??= []).push(...vals);
  }
  return { name, params, value };
}

/** First TYPE param, lowercased, as the Mailwoman `context` (e.g. `work`). */
function contextOf(prop: VCardProperty): string {
  const types = (prop.params['TYPE'] ?? []).map((t) => t.toLowerCase()).filter((t) => t !== 'pref');
  const first = types[0];
  if (first === undefined) return '';
  if (first === 'cell') return 'mobile';
  return first;
}

/** PREF weighting: vCard 4 `PREF=n`, or vCard 3 `TYPE=PREF` ⇒ 1. */
function prefOf(prop: VCardProperty): number {
  const p = prop.params['PREF']?.[0];
  if (p !== undefined) {
    const n = Number.parseInt(p, 10);
    if (Number.isFinite(n)) return n;
  }
  if ((prop.params['TYPE'] ?? []).some((t) => t.toLowerCase() === 'pref')) return 1;
  return 0;
}

function emptyName(): ContactName {
  return { full: '', given: '', surname: '', prefix: '', suffix: '' };
}

/** Parse a vCard document (possibly many cards) into contact drafts. */
export function parseVCards(text: string): ParsedContact[] {
  const lines = unfold(text);
  const cards: ParsedContact[] = [];
  let cur: ParsedContact | null = null;

  for (const line of lines) {
    const trimmed = line.trim();
    if (trimmed.length === 0) continue;
    const prop = parseLine(trimmed);
    if (prop === null) continue;

    if (prop.name === 'BEGIN' && prop.value.toUpperCase() === 'VCARD') {
      cur = blankContact();
      continue;
    }
    if (prop.name === 'END' && prop.value.toUpperCase() === 'VCARD') {
      if (cur !== null) cards.push(finalizeContact(cur));
      cur = null;
      continue;
    }
    if (cur === null) continue;
    applyProperty(cur, prop);
  }
  return cards;
}

function blankContact(): ParsedContact {
  return {
    uid: '',
    kind: 'individual',
    name: emptyName(),
    nicknames: [],
    organizations: [],
    titles: [],
    emails: [],
    phones: [],
    onlineServices: [],
    addresses: [],
    anniversaries: [],
    notes: '',
    photoBlobId: null,
    isFavorite: false,
    groupIds: [],
    pgpKey: null,
    smimeCert: null,
  };
}

function applyProperty(c: ParsedContact, prop: VCardProperty): void {
  const value = unescapeText(prop.value);
  switch (prop.name) {
    case 'UID':
      c.uid = value.replace(/^urn:uuid:/i, '');
      break;
    case 'KIND':
      c.kind = value.toLowerCase() === 'org' || value.toLowerCase() === 'organization' ? 'org' : 'individual';
      break;
    case 'FN':
      c.name.full = value;
      break;
    case 'N': {
      const [surname, given, , prefix, suffix] = splitValue(prop.value, ';').map(unescapeText);
      c.name.surname = surname ?? '';
      c.name.given = given ?? '';
      c.name.prefix = prefix ?? '';
      c.name.suffix = suffix ?? '';
      break;
    }
    case 'NICKNAME':
      c.nicknames.push(...splitValue(prop.value, ',').map(unescapeText).filter((n) => n.length > 0));
      break;
    case 'ORG':
      c.organizations.push(splitValue(prop.value, ';').map(unescapeText).filter((s) => s.length > 0).join(' · '));
      break;
    case 'TITLE':
    case 'ROLE':
      if (value.length > 0) c.titles.push(value);
      break;
    case 'EMAIL': {
      const email: ContactEmail = { context: contextOf(prop), value, pref: prefOf(prop) };
      c.emails.push(email);
      break;
    }
    case 'TEL': {
      const phone: ContactValue = { context: contextOf(prop), value };
      c.phones.push(phone);
      break;
    }
    case 'IMPP':
    case 'X-SOCIALPROFILE': {
      const svc: ContactValue = { context: contextOf(prop) || prop.name.replace(/^X-/, '').toLowerCase(), value };
      c.onlineServices.push(svc);
      break;
    }
    case 'ADR': {
      const [pobox, ext, street, locality, region, postcode, country] = splitValue(prop.value, ';').map(unescapeText);
      c.addresses.push({
        context: contextOf(prop),
        pobox: pobox ?? '',
        ext: ext ?? '',
        street: street ?? '',
        locality: locality ?? '',
        region: region ?? '',
        postcode: postcode ?? '',
        country: country ?? '',
      });
      break;
    }
    case 'BDAY':
      c.anniversaries.push({ kind: 'birthday', date: normalizeDate(value) });
      break;
    case 'ANNIVERSARY':
      c.anniversaries.push({ kind: 'anniversary', date: normalizeDate(value) });
      break;
    case 'NOTE':
      c.notes = c.notes.length > 0 ? `${c.notes}\n${value}` : value;
      break;
    case 'KEY':
      c.pgpKey = value;
      break;
    default:
      break;
  }
}

/** Coerce a vCard date (`19900101`, `1990-01-01`) to ISO `YYYY-MM-DD`. */
function normalizeDate(value: string): string {
  const compact = value.match(/^(\d{4})(\d{2})(\d{2})$/);
  if (compact) return `${compact[1]}-${compact[2]}-${compact[3]}`;
  return value;
}

function finalizeContact(c: ParsedContact): ParsedContact {
  // Derive a full name from the structured N when FN is absent.
  if (c.name.full.length === 0) {
    const parts = [c.name.prefix, c.name.given, c.name.surname, c.name.suffix].filter((p) => p.length > 0);
    c.name.full = parts.join(' ').trim();
  }
  if (c.name.full.length === 0 && c.organizations.length > 0) {
    c.name.full = c.organizations[0]!;
  }
  return c;
}

// ── Writing ──────────────────────────────────────────────────────────────────

function escapeText(value: string): string {
  return value.replace(/\\/g, '\\\\').replace(/\n/g, '\\n').replace(/,/g, '\\,').replace(/;/g, '\\;');
}

/** Fold a content line to <=75 chars per RFC 6350 (CRLF + leading space). */
function foldLine(line: string): string {
  if (line.length <= 75) return line;
  let out = line.slice(0, 75);
  let rest = line.slice(75);
  while (rest.length > 0) {
    out += `\r\n ${rest.slice(0, 74)}`;
    rest = rest.slice(74);
  }
  return out;
}

function emitProp(name: string, value: string, params: string[] = []): string {
  const head = params.length > 0 ? `${name};${params.join(';')}` : name;
  return foldLine(`${head}:${escapeText(value)}`);
}

/** Serialize one contact card as a vCard 4.0 entry. */
export function toVCard(card: ContactCard): string {
  const lines: string[] = ['BEGIN:VCARD', 'VERSION:4.0'];
  if (card.uid.length > 0) lines.push(emitProp('UID', card.uid));
  if (card.kind === 'org') lines.push(emitProp('KIND', 'org'));
  lines.push(emitProp('FN', card.name.full));
  lines.push(
    foldLine(
      `N:${[card.name.surname, card.name.given, '', card.name.prefix, card.name.suffix].map(escapeText).join(';')}`,
    ),
  );
  if (card.nicknames.length > 0) lines.push(emitProp('NICKNAME', card.nicknames.map(escapeText).join(',')));
  for (const org of card.organizations) lines.push(emitProp('ORG', org));
  for (const title of card.titles) lines.push(emitProp('TITLE', title));
  for (const email of card.emails) {
    const params: string[] = [];
    if (email.context.length > 0) params.push(`TYPE=${email.context}`);
    if (email.pref > 0) params.push(`PREF=${email.pref}`);
    lines.push(emitProp('EMAIL', email.value, params));
  }
  for (const phone of card.phones) {
    const params = phone.context.length > 0 ? [`TYPE=${phone.context}`] : [];
    lines.push(emitProp('TEL', phone.value, params));
  }
  for (const svc of card.onlineServices) {
    const params = svc.context.length > 0 ? [`TYPE=${svc.context}`] : [];
    lines.push(emitProp('IMPP', svc.value, params));
  }
  for (const anniv of card.anniversaries) {
    lines.push(emitProp(anniv.kind === 'birthday' ? 'BDAY' : 'ANNIVERSARY', anniv.date));
  }
  if (card.notes.length > 0) lines.push(emitProp('NOTE', card.notes));
  if (card.pgpKey !== null && card.pgpKey.length > 0) lines.push(emitProp('KEY', card.pgpKey));
  lines.push('END:VCARD');
  return lines.join('\r\n');
}

/** Serialize many cards as one vCard document. */
export function toVCardDocument(cards: ContactCard[]): string {
  return cards.map(toVCard).join('\r\n') + '\r\n';
}

/** Re-export helper so callers building anniversaries stay consistent. */
export function birthday(date: string): Anniversary {
  return { kind: 'birthday', date };
}

File: crates/mw-ics/src/vcard.rs
//! vCard 3/4 parse+emit against the frozen §2.1 `ContactCard` projection.
//!
//! Reading uses the `vcard4` crate (tolerant of both 3.0 and 4.0); writing is
//! hand-rolled over the vCard 4.0 grammar so `TYPE`/`PREF` parameters on
//! emails/phones survive the round-trip (the builder does not expose them).
//! `vcard_raw` is the round-trip source of truth (plan risk #13).

use serde_json::{Value, json};
use vcard4::parameter::Parameters;
use vcard4::parse_loose;
use vcard4::property::{Kind, Property as _, TextOrUriProperty};

use crate::{IcsError, Result};

/// One parsed contact: the `ContactCard` projection + verbatim `vcard_raw`.
#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
pub struct ParsedVcard {
    pub vcard_raw: String,
    pub json: Value,
}

fn ctx_from(params: Option<&Parameters>) -> String {
    params
        .and_then(|p| p.types.as_ref())
        .and_then(|t| t.first())
        .map(|t| t.to_string().to_ascii_lowercase())
        .unwrap_or_default()
}

fn pref_from(params: Option<&Parameters>) -> i64 {
    params.and_then(|p| p.pref).map(i64::from).unwrap_or(0)
}

fn tou_string(p: &TextOrUriProperty) -> String {
    p.to_string()
}

/// Parse a vCard 3/4 document into per-card Mailwoman projections.
pub fn parse_vcard(bytes: &[u8]) -> Result<Vec<ParsedVcard>> {
    let text = String::from_utf8_lossy(bytes);
    let cards = parse_loose(text.as_ref()).map_err(|e| IcsError::Vcard(format!("{e}")))?;
    let mut out = vec![];
    for card in cards {
        let full = card
            .formatted_name
            .first()
            .map(|p| p.value.clone())
            .unwrap_or_default();
        let (given, surname, prefix, suffix) = match &card.name {
            Some(n) => {
                let v = &n.value;
                (
                    v.get(1).cloned().unwrap_or_default(),
                    v.first().cloned().unwrap_or_default(),
                    v.get(3).cloned().unwrap_or_default(),
                    v.get(4).cloned().unwrap_or_default(),
                )
            }
            None => Default::default(),
        };
        let kind = match card.kind.as_ref().map(|k| &k.value) {
            Some(Kind::Org) | Some(Kind::Group) => "org",
            _ => "individual",
        };
        let emails: Vec<Value> = card
            .email
            .iter()
            .map(|e| {
                json!({
                    "context": ctx_from(e.parameters.as_ref()),
                    "value": e.value,
                    "pref": pref_from(e.parameters.as_ref()),
                })
            })
            .collect();
        let phones: Vec<Value> = card
            .tel
            .iter()
            .map(|t| json!({ "context": ctx_from(t.parameters()), "value": tou_string(t) }))
            .collect();
        let online: Vec<Value> = card
            .impp
            .iter()
            .map(|s| json!({ "context": ctx_from(s.parameters.as_ref()), "value": s.value.to_string() }))
            .collect();
        let organizations: Vec<Value> = card
            .org
            .iter()
            .map(|o| Value::String(o.value.join(";")))
            .collect();
        let titles: Vec<Value> = card
            .title
            .iter()
            .map(|t| Value::String(t.value.clone()))
            .collect();
        let nicknames: Vec<Value> = card
            .nickname
            .iter()
            .map(|n| Value::String(n.value.clone()))
            .collect();
        let notes = card
            .note
            .first()
            .map(|n| n.value.clone())
            .unwrap_or_default();
        let uid = card.uid.as_ref().map(tou_string).unwrap_or_default();
        let pgp_key = card
            .key
            .first()
            .map(tou_string)
            .map(Value::String)
            .unwrap_or(Value::Null);
        let mut anniversaries: Vec<Value> = vec![];
        if let Some(b) = &card.bday {
            anniversaries.push(json!({ "kind": "birthday", "date": b.to_string() }));
        }

        let json = json!({
            "id": uid,
            "addressBookId": "",
            "uid": uid,
            "kind": kind,
            "name": {
                "full": full, "given": given, "surname": surname,
                "prefix": prefix, "suffix": suffix,
            },
            "nicknames": nicknames,
            "organizations": organizations,
            "titles": titles,
            "emails": emails,
            "phones": phones,
            "onlineServices": online,
            "addresses": [],
            "anniversaries": anniversaries,
            "notes": notes,
            "photoBlobId": Value::Null,
            "isFavorite": false,
            "groupIds": [],
            "pgpKey": pgp_key,
            "smimeCert": Value::Null,
            "etag": Value::Null,
        });
        out.push(ParsedVcard {
            vcard_raw: card.to_string(),
            json,
        });
    }
    Ok(out)
}

// ── emit ─────────────────────────────────────────────────────────────────────

fn esc(s: &str) -> String {
    s.replace('\\', "\\\\")
        .replace(',', "\\,")
        .replace(';', "\\;")
        .replace('\n', "\\n")
}

fn s(v: &Value, k: &str) -> String {
    v.get(k)
        .and_then(Value::as_str)
        .unwrap_or_default()
        .to_string()
}

fn line(out: &mut String, name: &str, params: &str, value: &str) {
    out.push_str(name);
    out.push_str(params);
    out.push(':');
    out.push_str(value);
    out.push_str("\r\n");
}

/// Emit a vCard 4.0 document from a `ContactCard` projection.
pub fn emit_vcard(contact_json: &Value) -> Result<String> {
    let v = contact_json;
    let mut out = String::from("BEGIN:VCARD\r\nVERSION:4.0\r\n");

    let uid = s(v, "uid");
    if !uid.is_empty() {
        line(&mut out, "UID", "", &uid);
    }
    let kind = s(v, "kind");
    if !kind.is_empty() {
        line(&mut out, "KIND", "", &kind);
    }
    let name = v.get("name").cloned().unwrap_or(Value::Null);
    line(&mut out, "FN", "", &esc(&s(&name, "full")));
    let n = format!(
        "{};{};;{};{}",
        esc(&s(&name, "surname")),
        esc(&s(&name, "given")),
        esc(&s(&name, "prefix")),
        esc(&s(&name, "suffix")),
    );
    line(&mut out, "N", "", &n);

    for nick in v
        .get("nicknames")
        .and_then(Value::as_array)
        .into_iter()
        .flatten()
    {
        if let Some(t) = nick.as_str() {
            line(&mut out, "NICKNAME", "", &esc(t));
        }
    }
    for org in v
        .get("organizations")
        .and_then(Value::as_array)
        .into_iter()
        .flatten()
    {
        if let Some(t) = org.as_str() {
            // Structural `;` separators are preserved; components escaped on read.
            line(&mut out, "ORG", "", t);
        }
    }
    for title in v
        .get("titles")
        .and_then(Value::as_array)
        .into_iter()
        .flatten()
    {
        if let Some(t) = title.as_str() {
            line(&mut out, "TITLE", "", &esc(t));
        }
    }
    for email in v
        .get("emails")
        .and_then(Value::as_array)
        .into_iter()
        .flatten()
    {
        let mut params = String::new();
        let ctx = s(email, "context");
        if !ctx.is_empty() {
            params.push_str(&format!(";TYPE={ctx}"));
        }
        let pref = email.get("pref").and_then(Value::as_i64).unwrap_or(0);
        if pref > 0 {
            params.push_str(&format!(";PREF={pref}"));
        }
        line(&mut out, "EMAIL", &params, &s(email, "value"));
    }
    for phone in v
        .get("phones")
        .and_then(Value::as_array)
        .into_iter()
        .flatten()
    {
        let ctx = s(phone, "context");
        let params = if ctx.is_empty() {
            String::new()
        } else {
            format!(";TYPE={ctx}")
        };
        line(&mut out, "TEL", &params, &s(phone, "value"));
    }
    for svc in v
        .get("onlineServices")
        .and_then(Value::as_array)
        .into_iter()
        .flatten()
    {
        let ctx = s(svc, "context");
        let params = if ctx.is_empty() {
            String::new()
        } else {
            format!(";TYPE={ctx}")
        };
        line(&mut out, "IMPP", &params, &s(svc, "value"));
    }
    let notes = s(v, "notes");
    if !notes.is_empty() {
        line(&mut out, "NOTE", "", &esc(&notes));
    }
    if let Some(key) = v.get("pgpKey").and_then(Value::as_str) {
        line(&mut out, "KEY", "", key);
    }

    out.push_str("END:VCARD\r\n");
    Ok(out)
}
